### DISCUSSION TITLE
Adding alpha-porosity source term to the k-epsilon turbulence model

### DISCUSSION THREAD

**Discussion by chairmanmao256** [mdolab/dafoam#254] [General]
Hi! I'm here again! Recently I'm working on Topology Optimization of turbulent flow. To take the solid material into account in the turbulence model, I add alpha-porosity source terms to the model equations. They are just like the alpha-porosity source terms in the momentum equation. The exact form of the source terms can be found in the `.md` file attached.

[equation.md](https://github.com/mdolab/dafoam/files/7863288/equation.md)

It has been tested that the primal solution converges nicely. However, the adjoint solution does not converge. (The initial residual is about 1e-3. After 3000 GMRES iterations, it is still 1e-3) If I turn off the source term in the turbulence model, then the adjoint can converge. I've tried the solution listed on this [page](https://dafoam.github.io/mydoc_get_started_faq.html), but it didn't work.

Is there any other method to help the adjoint equation to converge? (For example, modifying more parameters related to the GMRES method, changing the method of solving linear equations or using other pre-condition methods ) Or has my modification caused the adjoint equation to be unsolvable?

Thank you!!!! Chinese new year is coming, @friedenhe 给您拜个早年！(Maybe too early)

**Comment by friedenhe** [comment]
Happy New Year! 
I can not see equation.md, can you do a screenshot of the terms you added to the KE model? Also, what have you changed in the source code and which case you are simulating? Did the adjoint fail even for the baseline design with the default alpha values, or the adjoint failed during the optimization when the alpha values were changed by the optimizer? Anyway, more details will be helpful to debug the case.

**Reply by chairmanmao256** [reply]
Thank you for your timely response! The following codes are added to `DAkEpsilon.C`. All the field variables are read from directory `0` in `createFieldsSimple.C`.
```C++
    const volScalarField& alphaPorosity_ = mesh_.thisDb().lookupObject<volScalarField>("alphaPorosity");
    const volScalarField& alphaMax_ = mesh_.thisDb().lookupObject<volScalarField>("alphaMax");
    const volScalarField& q1_ = mesh_.thisDb().lookupObject<volScalarField>("q1");
    const volScalarField& q2_ = mesh_.thisDb().lookupObject<volScalarField>("q2");
    const volScalarField& turFac_ = mesh_.thisDb().lookupObject<volScalarField>("turFac");
    //alphaPorosity is the design variable of Topology Optimization. Other fields are all constants.
```
I also modified the definition of the epsilon equation and the k equation in `DAkEpsilon.C`
```
    // Dissipation equation
    tmp<fvScalarMatrix> epsEqn(
        fvm::ddt(phase_, rho_, epsilon_)
            + fvm::div(phaseRhoPhi_, epsilon_)
            - fvm::laplacian(phase_ * rho_ * DepsilonEff(), epsilon_)
        == C1_ * phase_() * rho_() * G * epsilon_() / k_()
            - fvm::SuSp((scalar(2.0 / 3.0) * C1_ - C3_) * phase_() * rho_() * divU, epsilon_)
            - fvm::Sp(C2_ * phase_() * rho_() * epsilon_() / k_(), epsilon_)
            
            // source term starts
            - fvm::Sp(volScalarField("penaltye", turFac_ * q1_ * alphaPorosity_ / (q2_ - alphaPorosity_ / alphaMax_)), epsilon_) 
            // source term ends

            + epsilonSource());
```
```
    // Turbulent kinetic energy equation
    tmp<fvScalarMatrix> kEqn(
        fvm::ddt(phase_, rho_, k_)
            + fvm::div(phaseRhoPhi_, k_, divKScheme)
            - fvm::laplacian(phase_ * rho_ * DkEff(), k_)
        == phase_() * rho_() * G
            - fvm::SuSp((2.0 / 3.0) * phase_() * rho_() * divU, k_)
            - fvm::Sp(phase_() * rho_() * epsilon_() / k_(), k_)
                   
            // source term starts
            - fvm::Sp(volScalarField("penaltyk", turFac_ * q1_ * alphaPorosity_ / (q2_ - alphaPorosity_ / alphaMax_)), k_)
             // source term ends

            + kSource());
```
The analytical form of the source term is:
![equations](https://user-images.githubusercontent.com/68281328/149461237-36e19db5-40dd-4ad8-b2cb-6c7f767a3607.png)
The variable `turFac` in the codes stands for the parameter beta in the above equation.
Then I used my modified k-Epsilon model with the source term turned on (set `turFac` to 1) to simulate the following case. Non-uniform initial distribution of `alphaPorosity` defined by myself is implemented. The objective function is the drag of the porous media (also defined by myself). And the constraint is the volume of the porous media.
![initial_profile](https://user-images.githubusercontent.com/68281328/149458520-d142f5db-ca08-4879-9ddc-40690f615f7a.jpeg)
![initial_velocity](https://user-images.githubusercontent.com/68281328/149458538-435346e4-9cd2-4991-a2a7-7217c17dae04.jpeg)
The above two figures show the initial conditions and the initial flow field respectively. CFD converges for the baseline design, but the adjoint failed for the initial distribution of `alphaPorosity` (the baseline design). 
Then I turned off the source term in k-Epsilon model (set `turFac` to 0). The adjoint converged. Then I set a much smaller source term in the turbulence models(set `turFac` to 0.01). The adjoint converged but in a much slower way. (It takes more than 2000 iterations to drive the residual from 1e-3 to 1e-9) At last, I set `turFac` to 0.1, the residual remained 1e-3 even after 3500 iterations.
I also tried to add similar source terms to the k-Omega model. I used the modified k-Omega model to run the pitzDaily case. The adjoint converges for the baseline design. However, as the area where the `alphaPorosity` is larger than 0 increases, the convergence of adjoint becomes slower and slower. After about 5 adjoint evaluations, the adjoint's residual drops by only two orders of magnitude after 2000 iterations.
Combining the above results, the convergence rate of the adjoint equation should be related to the value of the source term in the turbulence model. The larger the source term is, the harder the adjoint equation converges. Is there any way to make the adjoint converge even when the source term is large?

**Reply by chairmanmao256** [reply]
The case file of the 2 figures mentioned above is uploaded here.
[githubDisccussion.zip](https://github.com/mdolab/dafoam/files/7868700/githubDisccussion.zip)

**Reply by friedenhe** [reply]
You forgot to add this to daOption: `"adjStateOrdering": "cell",`. Also, you need to give k and epsilon proper scalings in `normalizeStates` (check the NACA0012 tutorials). Let me know if these help.

**Reply by chairmanmao256** [reply]
Thank you for your suggestions! I've added the `"adjStateOrdering": "cell",` option and I normalize k and epsilon as follows:
```
"normalizeStates":{
    "U":U0,
    "p": U0 * U0/2.0,
    "nut": nuTilda0 * 10.0,
    "phi": 1.0,
    "k": k0 * 1e-3,
    "epsilon": epsilon0 * 1e-3
}
```
where `U0 = 0.26`, `k0 = 0.015`, 'epsilon0 = 0.07'. And `turFac` is set to 0.1.  The adjoint options are:
"adjEqnOption": {"gmresRelTol": 1.0e-8, "gmresMaxIters": 3500, "gmresRestart": 3500, "pcFillLevel": 2, "jacMatReOrdering": "natural"}
The convergence history at ''Evaluating Objective Function Sensitivities 001" is:
```
Main iteration 0 KSP Residual norm 1.153349686522e-03 195 s
Main iteration 100 KSP Residual norm 9.134714997637e-04 211 s
Main iteration 200 KSP Residual norm 3.359743138720e-04 229 s
Main iteration 300 KSP Residual norm 2.272713416652e-04 249 s
Main iteration 400 KSP Residual norm 1.992217875023e-04 271 s
Main iteration 500 KSP Residual norm 1.753769691460e-04 295 s
Main iteration 600 KSP Residual norm 1.580793986255e-04 322 s
Main iteration 700 KSP Residual norm 1.454712118674e-04 350 s
Main iteration 800 KSP Residual norm 1.397564784936e-04 381 s
Main iteration 900 KSP Residual norm 1.316058689916e-04 413 s
Main iteration 1000 KSP Residual norm 1.272301416057e-04 448 s
Main iteration 1100 KSP Residual norm 1.256907577885e-04 484 s
Main iteration 1200 KSP Residual norm 1.249090166410e-04 523 s
Main iteration 1300 KSP Residual norm 1.173993098334e-04 565 s
Main iteration 1400 KSP Residual norm 1.040914821472e-04 609 s
Main iteration 1500 KSP Residual norm 9.194836556461e-05 654 s
Main iteration 1600 KSP Residual norm 6.557788807997e-05 701 s
Main iteration 1700 KSP Residual norm 3.753723913001e-05 750 s
Main iteration 1800 KSP Residual norm 2.211751066328e-05 801 s
Main iteration 1900 KSP Residual norm 1.492636043371e-05 854 s
Main iteration 2000 KSP Residual norm 8.485294788010e-06 909 s
Main iteration 2100 KSP Residual norm 4.677655367822e-06 966 s
Main iteration 2200 KSP Residual norm 3.541451942757e-06 1026 s
Main iteration 2300 KSP Residual norm 2.837155492454e-06 1087 s
Main iteration 2400 KSP Residual norm 2.486346424107e-06 1151 s
Main iteration 2500 KSP Residual norm 1.777478784126e-06 1216 s
Main iteration 2600 KSP Residual norm 1.101798939812e-06 1283 s
Main iteration 2700 KSP Residual norm 8.837478844466e-07 1353 s
Main iteration 2800 KSP Residual norm 7.380420102947e-07 1425 s
Main iteration 2900 KSP Residual norm 4.674348206219e-07 1498 s
Main iteration 3000 KSP Residual norm 4.005658836672e-07 1575 s
Main iteration 3100 KSP Residual norm 2.264654931438e-07 1653 s
Main iteration 3200 KSP Residual norm 1.900170066319e-07 1738 s
Main iteration 3300 KSP Residual norm 1.453747664756e-07 1824 s
Main iteration 3400 KSP Residual norm 1.066776739068e-07 1908 s
Main iteration 3500 KSP Residual norm 7.257828317994e-08 1994 s
Main iteration 3500 KSP Residual norm 7.257828317994e-08 1995 s 
Total iterations 3500
```
It is much faster than my previous trial. But it still takes DAFoam half an hour to drive the residual from 1e-3 to 1e-8. Is there any way to make the adjoint converge even faster? 

Another question: Can I customize a solver that **uses turbulence model** when it solves primal but **exclude turbulence variables** (nuTilda; k and epsilon) when it solves adjoint? ('Frozen turbulence' assumption used by Topology Optimization community. I hope by using this assumption, the adjoint can converge faster. (Ｔ＿Ｔ）) For this solver, the dRdW might look like (the turbulence variables are omitted):
![dRdW](https://user-images.githubusercontent.com/68281328/149622940-8b8c0da6-1a0a-42c6-84ca-1eab51122010.png)
Maybe just copy the DASimpleFoam solver, then change the `stateInfo_` and `stateResConInfo_` of the copied solver (remove the turbulence variables from them)? Or maybe some other treatments need to be done?

Thank you in advance!

**Reply by friedenhe** [reply]
You may be able to remove the turbulence variable for the adjoint, but we do not suggest you do that. The frozen turbulence adjoint is not accurate and you will probably have optimization convergence issues. Two more things to try: 1. Use k0 and epsilon0 in `normalizeStates`, instead of using k0*1e-3. 2. Change `"pcFillLevel": 2` to `"pcFillLevel": 1`.

**Reply by chairmanmao256** [reply]
Thank you! Indeed, the frozen turbulence adjoint cannot give an accurate gradient and might affect the convergence of optimization. I'll try your suggestions and let you know when I have the results!

**Reply by chairmanmao256** [reply]
Thank you for your suggestion! After my holiday, I've tried to use k0 and epsilon0 in `normalizeStates`, and also use `"pcFillLevel":1 `, the convergence rate is still slow. However, when I use `k0*1e-4` in `normalizeStates`, although it converges slowly, the residual drops to 1e-8 after about 3000 iterations. So at least the current optimization problem is solved, but the reason for the slow convergence rate is still unclear.

**Reply by friedenhe** [reply]
Good to know that it converged better by tweaking the `normalStates`. You don't need to converge 8 orders of magnitude, usually 5 to 6 should be enough. Please keep us posted. Thanks!

**Reply by chairmanmao256** [reply]
Roger! Thank you for your suggestion!

**Reply by uttamcadambi07** [reply]
Hi I know this is an old thread but perhaps you could help me with this. So I have added a source term to my turbulence transport equation of the Spalart-Allmaras turbulence model. It is added as an explicit soruce. When I set the normalize factor for my nuTilda as nuTilda * 1e-04, I get very high sensitivities. My case is the periodic hill, with freestream velocity initialzied to 0.028. My bounds on the explicit source term are -1e-04 and 1e-04 with an initial value of 0. The initial case converges for both primal and adjoint solvers.  But the first optimization iteration, the adjoint solver fails. Much like the same way it has for you. I presume it is because of the normalize states, but any suggestions on how I can improve the case? Should I try an implicit source term perhaps? I must also mention that my dual infeasibility is quite high (100) to begin with.
